modules/frontend/tagsharding.go

```diff
@@ -0,0 +1,489 @@
+package frontend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/tempo/pkg/util"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/jsonpb" //nolint:all deprecated
+	"github.com/grafana/dskit/user"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/pkg/api"
+	"github.com/grafana/tempo/pkg/boundedwaitgroup"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/grafana/tempo/tempodb"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/opentracing/opentracing-go"
+)
+
+// shardedSearchResults is the overall response from the shardedSearchProgress
+type shardedSearchTagResults struct {
+	response         []string
+	statusCode       int
+	statusMsg        string
+	err              error
+	finishedRequests int
+}
+
+type shardedTagsSearchProgress interface {
+	setStatus(statusCode int, statusMsg string)
+	setError(err error)
+	addResponse(res []string)
+	shouldQuit() bool
+	result() *shardedSearchTagResults
+}
+
+type TagShardHandler interface {
+	unmarshalTagQuerierResponse(r io.ReadCloser) ([]string, error)
+	mashalResponseForFrontendQueier(valuetags []string) (string, error)
+}
+
+type TagValueHandler struct{}
+
+type TagNameHandler struct{}
+
+func (t *TagValueHandler) unmarshalTagQuerierResponse(r io.ReadCloser) ([]string, error) {
+	results := &tempopb.SearchTagValuesResponse{}
+	err := (&jsonpb.Unmarshaler{AllowUnknownFields: true}).Unmarshal(r, results)
+	if err != nil {
+		return nil, err
+	}
+	return results.TagValues, nil
+}
+
+func (t *TagValueHandler) mashalResponseForFrontendQueier(valuetags []string) (string, error) {
+	m := &jsonpb.Marshaler{}
+	bodyString, err := m.MarshalToString(&tempopb.SearchTagValuesResponse{
+		TagValues: valuetags,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return bodyString, nil
+}
```

Comment:
This only covers the V1 endpoint. `SearchTagValuesV2` should be supported as well.

Reply:
I'll add V2 endpoint also, just wanted to make sure this approach for V1 works first.